Orange/widgets/unsupervised/owtsne.py

```diff
@@ -1,60 +1,235 @@
+            task.data, task.pca_components, task.normalize
+        )
+        # Apply t-SNE's preprocessors to the data
+        task.pca_projection = task.tsne.preprocess(pca_projection)
+        state.set_partial_result(("pca_projection", task))
+
+    @staticmethod
+    def compute_initialization(task, state, **_):
+        # Prepare initial positions for t-SNE
+        state.set_status("Preparing initialization...")
+        task.initialization = task.tsne.compute_initialization(task.pca_projection.X)
+        state.set_partial_result(("initialization", task))
+
+    @staticmethod
+    def compute_affinities(task, state, **_):
+        # Compute affinities
+        state.set_status("Finding nearest neighbors...")
+        task.affinities = task.tsne.compute_affinities(task.pca_projection.X)
+        state.set_partial_result(("affinities", task))
+
+    @staticmethod
+    def compute_tsne(task, state, progress_callback=None):
+        tsne = task.tsne
+
+        state.set_status("Running optimization...")
+
+        # If this the first time we're computing t-SNE (otherwise we may just
+        # be resuming optimization), we have to assemble the tsne object
+        if task.tsne_embedding is None:
+            # Assemble a t-SNE embedding object and convert it to a TSNEModel
+            task.tsne_embedding = tsne.prepare_embedding(
+                task.affinities, task.initialization
+            )
+            task.tsne_embedding = tsne.convert_embedding_to_model(
+                task.pca_projection, task.tsne_embedding
+            )
+            state.set_partial_result(("tsne_embedding", task))
 
-        # Larger data sets need a larger number of iterations
-        if self.n_samples > 100_000:
-            self.early_exagg_iter, self.n_iter = 500, 1000
-        else:
-            self.early_exagg_iter, self.n_iter = 250, 750
+            if state.is_interruption_requested():
+                return
+
+        total_iterations_needed = tsne.early_exaggeration_iter + tsne.n_iter
+
+        # Run early exaggeration phase
+        while task.iterations_done < tsne.early_exaggeration_iter:
+            # Step size can't be larger than the remaining number of iterations
+            step_size = min(
+                _STEP_SIZE, tsne.early_exaggeration_iter - task.iterations_done
+            )
+            task.tsne_embedding = task.tsne_embedding.optimize(
+                step_size,
+                exaggeration=tsne.early_exaggeration,
+                momentum=0.5,
+                inplace=False,
+            )
+            task.iterations_done += step_size
+            state.set_partial_result(("tsne_embedding", task))
+            if progress_callback is not None:
+                progress_callback(task.iterations_done / total_iterations_needed)
+
+            if state.is_interruption_requested():
+                return
+
+        # Run regular optimization phase
+        while task.iterations_done < total_iterations_needed:
```

Comment:
This code is duplicated (with the exception of some parameters). Consider writing a function and calling it twice with different parameters.

Reply:
Done.